include/lbann/utils/serialization/serialize_matrices.hpp

```diff
@@ -0,0 +1,402 @@
+ *           matrix. Serializing views is not supported, and, in the
+ *           context of LBANN, should be unnecessary as the views will
+ *           be reestablished when setup() is called on the
+ *           deserialized objects.
+ *
+ *  @tparam ArchiveT (Inferred) The Cereal archive type to use.
+ *  @tparam T (Inferred) The data type of the matrix.
+ *  @param archive The Cereal archive into which the matrix will be written.
+ *  @param mat The matrix to serialize.
+ *  @throws lbann::exception Thrown when the matrix is actually a view.
+ *  @ingroup serialization
+ */
+template <typename ArchiveT, typename T,
+          lbann::utils::WhenNotTextArchive<ArchiveT> = 1>
+void save(ArchiveT& archive, El::Matrix<T, El::Device::GPU> const& mat)
+{
+  LBANN_ASSERT(!mat.Viewing());
+  El::Matrix<T, El::Device::CPU> cpu_mat(mat);
+  archive(cpu_mat);
+}
+#endif // defined LBANN_HAS_GPU
+
+/** @brief "Load" a CPU Matrix from a text-based archive.
+ *
+ *  The "text-based" ("human-readable") archives (XML and JSON) will
+ *  only contain basic matrix metadata (height and width). Thus, a
+ *  true deserialization will not be possible. Instead, the matrix
+ *  restored by one of these archive types will have the proper size
+ *  but it will not have meaningful data. Since these archive types
+ *  are primarily intended for debugging, this should not be a
+ *  problem. If it is, open an issue and it will be remedied.
+ *
+ *  @tparam ArchiveT (Inferred) The Cereal archive type to use.
+ *  @tparam T (Inferred) The data type of the matrix.
+ *  @param archive The Cereal archive from which the matrix will be read.
+ *  @param mat The target matrix to deserialize.
+ *  @throws lbann::exception The input matrix is already setup as a view.
+ *  @todo Perhaps it's better to throw an exception for these archives?
+ *  @ingroup serialization
+ */
+template <typename ArchiveT, typename T, El::Device D,
+          lbann::utils::WhenTextArchive<ArchiveT> = 1>
+void load(ArchiveT& archive, El::Matrix<T, D>& mat)
+{
+  LBANN_ASSERT(!mat.Viewing());
+  El::Int height, width;
+  archive(height, width);
+  mat.Resize(height, width);
+}
+
+/** @brief Load a CPU Matrix from a non-text archive.
+ *
+ *  @tparam ArchiveT (Inferred) The Cereal archive type to use.
+ *  @tparam T (Inferred) The data type of the matrix.
+ *  @param archive The Cereal archive from which the matrix will be read.
+ *  @param mat The target matrix to deserialize.
+ *  @throws lbann::exception The input matrix is already setup as a view.
+ *  @ingroup serialization
+ */
+template <typename ArchiveT, typename T,
+          lbann::utils::WhenNotTextArchive<ArchiveT> = 1>
+void load(ArchiveT& archive,
+          El::Matrix<T, El::Device::CPU>& mat)
+{
+  LBANN_ASSERT(!mat.Viewing());
+  El::Int height, width;
+  archive(height, width);
+  mat.Resize(height, width);
+  archive(binary_data(mat.Buffer(),
+                      mat.Height()*mat.Width()*sizeof(T)));
```

Comment:
Yes, see the test. I suppose I should be a bit more careful, since there's an implicit assumption that on load, `mat.Height() == mat.LDim()`. The `Resize` should take care of this in current Hydrogen, but I don't actually like that either.